### DISCUSSION TITLE
exception 9

### DISCUSSION THREAD

**Discussion by zeehond1** [esp8266/Arduino#8160] [General]
After upgrade from 2.7.4 to 3.0.0
 this small program crashes:
````
class test3 
  {
public:
  test3(int);
  ~test3();
  
  void body(void);
  int  body(void *mp);
  int  body(int mp);
  };

test3::test3(int axis)
  {
  Serial.println("Constructor test::test");
  }

test3::~test3(void)
  {
  Serial.println("destructor test::~test");
  }

void test3::body(void)
  {
  Serial.println("body(void)");
  }

int test3::body(void *mp)
  {
  Serial.println("body &mp");
  }


int test3::body(int mp)
  {
  Serial.println((const char *)"body (int)mp");
  }

void setup() {
  test3 *cl;
  
  Serial.begin(115200);
  while (!Serial);

  Serial.println("\n\n\n");
  Serial.println("Start:");
  cl=new test3(1);
  Serial.println(".");
  cl->body();
  Serial.println(".");
  cl->body(25);
  Serial.println(".");
  cl->body((void *)0);
  Serial.println(".");
  delete cl;
  Serial.println("Finished...");
  }

void loop() {
}
````
Output (snip):
````
20:11:46.005 -> 
20:11:46.005 -> 
20:11:46.005 -> 
20:11:46.005 -> 
20:11:46.005 -> Start:
20:11:46.005 -> Constructor test::test
20:11:46.005 -> .
20:11:46.005 -> body(void)
20:11:46.005 -> .
20:11:46.005 -> body (int)mp
20:11:46.005 -> 
20:11:46.005 -> --------------- CUT HERE FOR EXCEPTION DECODER ---------------
20:11:46.005 -> 
20:11:46.005 -> Exception (9):
20:11:46.005 -> epc1=0x402010b0 epc2=0x00000000 epc3=0x00000000 excvaddr=0x00000026 depc=0x00000000
20:11:46.039 -> 
20:11:46.039 -> >>>stack>>>
20:11:46.039 -> 
20:11:46.039 -> ctx: cont
20:11:46.039 -> sp: 3ffffe00 end: 3fffffc0 offset: 0190
20:11:46.039 -> 3fffff90:  3fffdad0 3ffe87eb 3ffee4f8 402010b0  
20:11:46.039 -> 3fffffa0:  3fffdad0 00000000 3ffee54c 402019d0  
20:11:46.039 -> 3fffffb0:  feefeffe feefeffe 3ffe85d8 40100b55  
20:11:46.039 -> <<<stack<<<
20:11:46.039 -> 
20:11:46.039 -> --------------- CUT HERE FOR EXCEPTION DECODER ---------------
````
Expected output (with 2.7.4)
````
20:26:36.418 -> 
20:26:36.418 -> 
20:26:36.418 -> 
20:26:36.418 -> Start:
20:26:36.418 -> Constructor test::test
20:26:36.418 -> .
20:26:36.418 -> body(void)
20:26:36.418 -> .
20:26:36.418 -> body (int)mp
20:26:36.418 -> .
20:26:36.418 -> body &mp
20:26:36.418 -> .
20:26:36.418 -> destructor test::~test
20:26:36.418 -> Finished...
````

Is running fine on earlier releases, ofcourse. I can not see the relation with a misaligned pointer, of see what the release dependent difference can be, does someone? 
Selected board is generic esp8266, upload output:
Executable segment sizes:
IROM   : 233304          - code in flash         (default or ICACHE_FLASH_ATTR) 
IRAM   : 26888   / 32768 - code in IRAM          (ICACHE_RAM_ATTR, ISRs...) 
DATA   : 1248  )         - initialized variables (global, static) in RAM/HEAP 
RODATA : 764   ) / 81920 - constants             (global, static) in RAM/HEAP 
BSS    : 24936 )         - zeroed variables      (global, static) in RAM/HEAP 
Sketch uses 262204 bytes (27%) of program storage space. Maximum is 958448 bytes.
Global variables use 26948 bytes (32%) of dynamic memory, leaving 54972 bytes for local variables. Maximum is 81920 bytes.
/home/ed/.arduino15/packages/esp8266/tools/python3/3.7.2-post1/python3 /home/ed/.arduino15/packages/esp8266/hardware/esp8266/2.7.4/tools/upload.py --chip esp8266 --port /dev/ttyUSB0 --baud 115200 --before default_reset --after hard_reset write_flash 0x0 /tmp/arduino_build_599182/nodemcu_classtest.ino.bin 
esptool.py v2.8
Serial port /dev/ttyUSB0
Connecting....
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: cc:50:e3:c6:93:20
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Auto-detected Flash size: 4MB
Flash params set to 0x034f
Compressed 266352 bytes to 196309...
Wrote 266352 bytes (196309 compressed) at 0x00000000 in 17.3 seconds (effective 123.4 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...

I tried several computers. Can someone confirm this problem?

**Comment by earlephilhower** [comment]
Thanks for the MCVE.  I am able to reproduce it locally.  What's really weird, though, is that I tried to get better debugging by using `-O0` (i.e. no GCC optimization) and it passes there.  Only `-Os` seem to have the fault...

## -O0
````
Start:
Constructor test::test
.
body(void)
.
body (int)mp
.
body &mp
.
destructor test::~test
Finished...
````

## - Os (normal Arduino build)
````

Start:
Constructor test::test
.
body(void)
.
body (int)mp

--------------- CUT HERE FOR EXCEPTION DECODER ---------------

Exception (9):
epc1=0x402010b0 epc2=0x00000000 epc3=0x00000000 excvaddr=0x00000026 depc=0x00000000

>>>stack>>>

ctx: cont
sp: 3ffffe00 end: 3fffffc0 offset: 0190
3fffff90:  3fffdad0 3ffe87eb 3ffee4f8 402010b0  
3fffffa0:  3fffdad0 00000000 3ffee54c 402019d0  
3fffffb0:  feefeffe feefeffe 3ffe85d8 40100b55  
<<<stack<<<

--------------- CUT HERE FOR EXCEPTION DECODER ---------------
````

Decoded:
````
Exception 9: LoadStoreAlignmentCause: Load or store to an unaligned address
PC: 0x402010b0: HardwareSerial::available() at /home/earle/Arduino/hardware/esp8266com/esp8266/cores/esp8266/HardwareSerial.cpp line 110
EXCVADDR: 0x00000026

Decoding stack results
0x402010b0: HardwareSerial::available() at /home/earle/Arduino/hardware/esp8266com/esp8266/cores/esp8266/HardwareSerial.cpp line 110
0x402019d0: loop_wrapper() at /home/earle/Arduino/hardware/esp8266com/esp8266/cores/esp8266/core_esp8266_main.cpp line 198
````

https://github.com/esp8266/Arduino/blob/d283541380b776956edd67ec39d849704645e632/cores/esp8266/HardwareSerial.cpp#L108-L115


https://github.com/esp8266/Arduino/blob/d283541380b776956edd67ec39d849704645e632/cores/esp8266/uart.cpp#L218-L229

**Reply by mcspr** [reply]
Was just typing a reply... Notice there is no `return something;` in the methods. It is technically undefined behaviour, but is allowed *and* also shown as a mere warning (also see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=43943)
```
blah.cpp: In member function ‘int Test::method(void*)’:
blah.cpp:10:5: warning: no return statement in function returning non-void [-Wreturn-type]
   10 |     }
      |     ^
.⏎
```

**Reply by earlephilhower** [reply]
Yes, I just noticed that too.

Adding the proper `return 0;` to the methods and all works.  So, this seems to be a real bug in the MCVE code and not something in the core.

**Reply by earlephilhower** [reply]
With `return 0;` in the 2 `int` returning methods

````
Start:
Constructor test::test
.
body(void)
.
body (int)mp
.
body &mp
.
destructor test::~test
Finished...
````

**Comment by mhightower83** [comment]
I was poking at this one also. I have been noticing lately that when I do some quick coding and forget the return, this kind of crash occurs. I didn't think too much about it, though I was surprised. I think, the resulting compile for a missing return use to work.  Looks like the optimizer is a lot more aggressive in v10. There have been other optimizations that have bitten me. I wish I had kept track of them.

**Reply by mcspr** [reply]
New standards also allow to not specify the type, and use either `auto` (similar to variable declaration, `&` and `const` can be around it) or `decltype(auto)` (standalone, can be used with templates)
```cpp
struct Test {
  auto method(void*) {
    std::cout << "method(void*)" << std::endl;
  }
};
```

**Comment by hreintke** [comment]
If a crash (almost) always occurs when no return statement is used you can consider `Werror=return-type` being set by default.
Then the user is always triggered and forced to solve the root cause.

**Reply by zeehond1** [reply]
All thanks for the help and the advises.
As the Arduino IDE does not give a warning or error, I have to be alert my selve. 
Ed.
